x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwkSetLoader.java

```diff
@@ -0,0 +1,201 @@
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class JwkSetLoader implements Releasable {
+
+    private static final Logger logger = LogManager.getLogger(JwkSetLoader.class);
+
+    private final AtomicReference<ListenableFuture<ContentAndJwksAlgs>> reloadFutureRef = new AtomicReference<>();
+    private final RealmConfig realmConfig;
+    private final List<String> allowedJwksAlgs;
+    private final String jwkSetPath;
+    @Nullable
+    private final URI jwkSetPathUri;
+    @Nullable
+    private final CloseableHttpAsyncClient httpClient;
+    private volatile ContentAndJwksAlgs contentAndJwksAlgs;
+
+    public JwkSetLoader(final RealmConfig realmConfig, final SSLService sslService) {
+        this.realmConfig = realmConfig;
+        this.allowedJwksAlgs = realmConfig.getSetting(JwtRealmSettings.ALLOWED_SIGNATURE_ALGORITHMS)
+            .stream()
+            .filter(JwtRealmSettings.SUPPORTED_SIGNATURE_ALGORITHMS_PKC::contains)
+            .toList();
+        this.jwkSetPath = realmConfig.getSetting(JwtRealmSettings.PKC_JWKSET_PATH);
+        assert Strings.hasText(this.jwkSetPath);
+        this.jwkSetPathUri = JwtUtil.parseHttpsUri(jwkSetPath);
+        if (this.jwkSetPathUri == null) {
+            this.httpClient = null;
+        } else {
+            this.httpClient = JwtUtil.createHttpClient(realmConfig, sslService);
+        }
+
+        // Any exception during loading requires closing JwkSetLoader's HTTP client to avoid a thread pool leak
+        try {
+            final PlainActionFuture<ContentAndJwksAlgs> future = new PlainActionFuture<>();
+            load(future);
+            // ASSUME: Blocking read operations are OK during startup
+            contentAndJwksAlgs = future.actionGet();
+        } catch (Throwable t) {
+            close();
+            throw t;
+        }
+    }
+
+    /**
+     * Load the JWK sets and pass its content to the specified listener.
+     */
+    void load(final ActionListener<ContentAndJwksAlgs> listener) {
+        final ListenableFuture<ContentAndJwksAlgs> future = this.getFuture();
+        future.addListener(listener);
+    }
+
+    /**
+     * Reload the JWK sets, compare to existing JWK sets and update it to the reloaded value if
+     * they are different. The listener is called with false if the reloaded content is the same
+     * as the existing one or true if they are different.
+     */
+    void reload(final ActionListener<Boolean> listener) {
+        load(ActionListener.wrap(newContentAndJwksAlgs -> {
+            if (Arrays.equals(contentAndJwksAlgs.sha256, newContentAndJwksAlgs.sha256)) {
+                // No change in JWKSet
+                listener.onResponse(false);
+            } else {
+                contentAndJwksAlgs = newContentAndJwksAlgs;
+                listener.onResponse(true);
+            }
+        }, listener::onFailure));
```

Comment:
Nice catch, thanks! I re-arranged the code so that field mutation is done inside `loadInternal` which is guaranteed to be single threaded. The listner now only returns a `boolean` to indicated whether the field is actually mutated because of updated JWK set. This simplifies the code a bit and allows us to remove the `load` method altogether.